lib/ansible/modules/cloud/amazon/ec2_vpc_igw.py

```diff
@@ -80,127 +83,181 @@
+            self._connection.delete_internet_gateway(InternetGatewayId=igw['internet_gateway_id'])
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            self._module.fail_json_aws(e, msg="Unable to delete Internet Gateway")
 
-    if check_mode:
-        return {'changed': True}
+        return self._results
 
-    try:
-        vpc_conn.detach_internet_gateway(igw.id, vpc_id)
-        vpc_conn.delete_internet_gateway(igw.id)
-    except EC2ResponseError as e:
-        raise AnsibleIGWException(
-            'Unable to delete Internet Gateway, error: {0}'.format(e))
+    def ensure_igw_present(self, vpc_id, tags):
+        self.check_input_tags(tags)
 
-    return {'changed': True}
+        igw = self.get_matching_igw(vpc_id)
 
+        if igw is None:
+            if self._check_mode:
+                self._results['changed'] = True
+                self._results['gateway_id'] = None
+                return self._results
 
-def ensure_igw_present(vpc_conn, vpc_id, tags, check_mode):
-    igw = get_matching_igw(vpc_conn, vpc_id)
-    changed = False
-    if igw is None:
-        if check_mode:
-            return {'changed': True, 'gateway_id': None}
+            try:
+                response = self._connection.create_internet_gateway()
+                igw = camel_dict_to_snake_dict(response['InternetGateway'])
+                self._connection.attach_internet_gateway(InternetGatewayId=igw['internet_gateway_id'], VpcId=vpc_id)
+                self._results['changed'] = True
+            except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+                self._module.fail_json_aws(e, msg='Unable to create Internet Gateway')
 
-        try:
-            igw = vpc_conn.create_internet_gateway()
-            vpc_conn.attach_internet_gateway(igw.id, vpc_id)
-            changed = True
-        except EC2ResponseError as e:
-            raise AnsibleIGWException(
-                'Unable to create Internet Gateway, error: {0}'.format(e))
+        igw['vpc_id'] = vpc_id
 
-    igw.vpc_id = vpc_id
+        igw['tags'] = self.ensure_tags(igw_id=igw['internet_gateway_id'], tags=tags, add_only=False)
 
-    if tags != igw.tags:
-        if check_mode:
-            check_mode_tags = ensure_tags(vpc_conn, igw.id, tags, False, check_mode)
-            igw_info = get_igw_info(igw)
-            igw_info.get('tags', {}).update(check_mode_tags.get('tags', {}))
-            return {'changed': True, 'gateway': igw_info}
-        ensure_tags(vpc_conn, igw.id, tags, False, check_mode)
-        igw.tags = tags
-        changed = True
+        igw_info = self.get_igw_info(igw)
+        self._results.update(igw_info)
 
-    igw_info = get_igw_info(igw)
+        return self._results
 
-    return {
-        'changed': changed,
-        'gateway': igw_info
-    }
+    def fail(self, **kwargs):
```

Comment:
I don't think this is really needed (and doesn't seem to be used)

Reply:
Ok